### PR TITLE
Trim down MCM tests.

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
MCM tests are slow and dominate the CI run time. New tests have been incorporated as new features came in without rationalizing the test suite as a whole. There are redundant or overly parametrized tests as a results.   

**Description of the Change:**
Combine a bunch of tests into `test_simple_dynamic_circuit` which tests all possible measurements are handle correctly by `dynamic_one_shot` post-processing. This needs not be parametrized over `reset` since it does not show up in post-processing.

Combine some tests into `test_multiple_measurements_and_reset` which checks that multiple measurements are handle correctly and that `reset` is properly applied while getting to the final state.

Combine all composite MCM measurements in `test_composite_mcms`.

`test_counts_return_type` remains as is. 

**Benefits:**
Faster CI.

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-65255]

### CODE REVIEW

File: tests/test_native_mcm.py [hunk 1]
```diff
@@ -230,181 +175,162 @@ def obs_tape(x, y, z, reset=False, postselect=None):
...
 
-@flaky(max_runs=5)
-@pytest.mark.parametrize("shots", [5000])
 @pytest.mark.parametrize("postselect", [None, 0, 1])
 @pytest.mark.parametrize("reset", [False, True])
-@pytest.mark.parametrize("measure_f", [qml.counts, qml.expval, qml.probs, qml.sample, qml.var])
-def test_single_mcm_multiple_measurements(shots, postselect, reset, measure_f):
-    """Tests that LightningDevice handles a circuit with a single mid-circuit measurement with reset
+def test_multiple_measurements_and_reset(postselect, reset):
+    """Tests that DefaultQubit handles a circuit with a single mid-circuit measurement with reset
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:222]:
```suggestion
    """Tests that LightningQubit handles a circuit with a single mid-circuit measurement with reset
```

File: tests/test_native_mcm.py [hunk 2]
```diff
@@ -230,181 +175,162 @@ def obs_tape(x, y, z, reset=False, postselect=None):
...
+        lambda x: 1 - x,
+        lambda x: x + 1,
+        lambda x: x & 3,
+        "mix",
+        "list",
+    ],
+)
+@pytest.mark.parametrize("measure_f", [qml.counts, qml.expval, qml.probs, qml.sample, qml.var])
+def test_composite_mcms(mcm_f, measure_f):
+    """Tests that DefaultQubit handles a circuit with a composite mid-circuit measurement and a
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:265]:
```suggestion
    """Tests that LightningQubit handles a circuit with a composite mid-circuit measurement and a
```

File: tests/test_native_mcm.py [hunk 3]
```diff
@@ -230,181 +175,162 @@ def obs_tape(x, y, z, reset=False, postselect=None):
...
+    [
+        lambda x, y: x + y,
+        lambda x, y: x - 7 * y,
+        lambda x, y: x & y,
+        lambda x, y: x == y,
+        lambda x, y: 4.0 * x + 2.0 * y,
+    ],
+)
+def test_counts_return_type(mcm_f):
+    """Tests that DefaultQubit returns the same keys for ``qml.counts`` measurements with ``dynamic_one_shot`` and ``defer_measurements``."""
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:317]:
```suggestion
    """Tests that LightningQubit returns the same keys for ``qml.counts`` measurements with ``dynamic_one_shot`` and ``defer_measurements``."""
```

**Review by AmintorDusko — COMMENTED** [review]
Looks good so far!

File: .github/workflows/tests_lkcpu_python.yml [hunk 4]
```diff
@@ -209,7 +209,7 @@ jobs:
           python -m pip install -r requirements-dev.txt
           python -m pip install openfermionpyscf
           if [ '${{ inputs.lightning-version }}' != 'stable' ]; then
-            SKIP_COMPILATION=True python -m pip install . -vv
+            python -m pip install . -vv
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply .github/workflows/tests_lkcpu_python.yml:212]: Some tests were complaining the LQubit binaries could not be found, so I built them.

File: tests/test_native_mcm.py [hunk 5]
```diff
@@ -29,6 +29,11 @@
     pytest.skip("No binary module found. Skipping.", allow_module_level=True)
 
 
+def get_device(wires, **kwargs):
+    kwargs.setdefault("shots", None)
+    return qml.device("lightning.qubit", wires=wires, **kwargs)
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply tests/test_native_mcm.py:34]: That must have been my mistake. Thanks!

**Review by AmintorDusko — APPROVED** [review]
Thank you @vincentmr. Great PR!

**Review by multiphaseCFD — APPROVED** [review]
Thanks @vincentmr ! Just a few question but happy to approve.

File: tests/test_native_mcm.py [hunk 6]
```diff
@@ -230,181 +175,162 @@ def obs_tape(x, y, z, reset=False, postselect=None):
...
+        return qml.counts(op=mcm_f(m0, m1))
 
-    func1 = qml.QNode(func, dev)
-    func2 = qml.defer_measurements(qml.QNode(func, dq))
-
-    results1 = func1(param)
-    results2 = func2(param)
-
-    validate_measurements(measure_f, shots, results1, results2)
+    results1 = qml.QNode(func, dev, mcm_method="one-shot")(param)
```
Comment (multiphaseCFD) [inline tests/test_native_mcm.py:333]: Can we pass `diff_method` to `qml.QNode` as well?

Comment (vincentmr) [reply tests/test_native_mcm.py:333]: Should not be an issue no, but not all diff methods are supported. Even parameter-shift isn't supported with post-selection not `None`. Anything in mind?

Comment (multiphaseCFD) [reply tests/test_native_mcm.py:333]: Nothing specifically. Just want to understand more about this feature

**Pull request merged**